users/src/test/java/io/crate/http/netty/HttpAuthUpstreamHandlerTest.java

```diff
@@ -0,0 +1,213 @@
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Matchers;
+
+import javax.annotation.Nullable;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.util.concurrent.CompletableFuture;
+
+import static org.elasticsearch.mock.orig.Mockito.times;
+import static org.hamcrest.core.Is.is;
+import static org.mockito.Mockito.*;
+
+public class HttpAuthUpstreamHandlerTest extends CrateUnitTest {
+
+    private InetSocketAddress IPv4_LOCALHOST = new InetSocketAddress(InetAddresses.forString("127.0.0.1"), 54321);
+
+    private final AuthenticationMethod denyAll = new AuthenticationMethod() {
+        @Override
+        public CompletableFuture<Boolean> pgAuthenticate(Channel channel, SessionContext session) {
+            return CompletableFutures.COMPLETED_FALSE;
+        }
+
+        @Override
+        public String name() {
+            return "denyAll";
+        }
+
+        @Override
+        public CompletableFuture<Boolean> httpAuthentication(String username) {
+            return CompletableFutures.COMPLETED_FALSE;
+        }
+    };
+
+    private final Authentication authService = new Authentication() {
+        @Override
+        public boolean enabled() {
+            return true;
+        }
+
+        @Nullable
+        @Override
+        public AuthenticationMethod resolveAuthenticationType(String user, InetAddress address, HbaProtocol protocol) {
+            // we want to test two things:
+            // 1) no authentication method
+            // 2) deny authentication
+            if ("null".equals(user)) {
+                return null;
+            }
+            return denyAll;
+        }
+    };
+
+    private ChannelHandlerContext getChannelHandlerContext(Channel ch) {
+        ChannelFuture cf = new DefaultChannelFuture(ch, false);
+        when(ch.write(Matchers.any())).thenReturn(cf);
+        when(ch.getRemoteAddress()).thenReturn(IPv4_LOCALHOST);
+
+        ChannelHandlerContext ctx = mock(ChannelHandlerContext.class);
+        when(ctx.getChannel()).thenReturn(ch);
+        when(ctx.canHandleUpstream()).thenReturn(true);
+        return ctx;
+    }
+
+    private void assertUnauthorized(Channel ch, @Nullable String error) {
+        ArgumentCaptor<HttpResponse> writeCaptor = ArgumentCaptor.forClass(HttpResponse.class);
+        verify(ch).write(writeCaptor.capture());
+        HttpResponse response = writeCaptor.getValue();
+        assertThat(response.getStatus(), is(HttpResponseStatus.UNAUTHORIZED));
+        ChannelBuffer content = error == null ? ChannelBuffers.EMPTY_BUFFER : ChannelBuffers.copiedBuffer(error, CharsetUtil.US_ASCII);
+        assertThat(response.getContent(), is(content));
+    }
+
+    @Test
+    public void testChannelClosedWhenUnauthorized() throws Exception {
+        Channel ch = mock(Channel.class);
+        ChannelFuture cf = new DefaultChannelFuture(ch, false);
+        when(ch.write(Matchers.any())).thenReturn(cf);
```

Comment:
I'll add some comments to make it more clear.
a lot of implementations in netty are private or package-private and only the interface is public. that makes it hard to use real instances.